### DISCUSSION TITLE
Arrow up - show first this session history (Help)

### DISCUSSION THREAD

**Discussion by mrbarletta** [nushell/nushell#8143] [Q&A]
Hi there, 

I digged for hours on how to approach this.  
I found someone posted a Menu that shows history from this session but didn't find a way for the Arrow UP to show the last command.

Is there a way to overwrite the default keybinding or re-implement the current one to filter first the session and then other commands?

Thanks!

**Comment by fdncred** [comment]
Did you see this? https://github.com/nushell/nu_scripts/tree/main/custom-menus

There's also something like this
```
  {
    name: fuzzy_history_fzf
    modifier: control
    keycode: char_r
    mode: [emacs , vi_normal, vi_insert]
    event: {
      send: executehostcommand
      cmd: "commandline (
        history 
          | each { |it| $it.command } 
          | uniq 
          | reverse 
          | str collect (char -i 0) 
          | fzf --read0 --tiebreak=chunk --layout=reverse  --multi --preview='echo {..}' --preview-window='bottom:3:wrap' --bind alt-up:preview-up,alt-down:preview-down --height=70% -q (commandline) 
          | decode utf-8 
          | str trim
      )"
    }
  }
```
What I'm not sure about is how to assign up or down arrow. I'm not sure that exists in reedline yet. We'd have to look.


I see this in the reedline code. I wonder if we can do `modifier: control` and `keycode: char_p` for up arrow?
```rust
    // EMACS arrows
    kb.add_binding(
        KM::CONTROL,
        KC::Char('p'),
        ReedlineEvent::UntilFound(vec![ReedlineEvent::MenuUp, ReedlineEvent::Up]),
    );
    kb.add_binding(
        KM::CONTROL,
        KC::Char('n'),
        ReedlineEvent::UntilFound(vec![ReedlineEvent::MenuDown, ReedlineEvent::Down]),
    );
```
What I mean, is assigning ctrl+p responds to the up arrow button?

**Comment by fdncred** [comment]
i just found this in the nushell code
```rust
        "down" => KeyCode::Down,
        "up" => KeyCode::Up,
```

So that means you should be able to do `modifier: none` and `keycode: up` in a keybinding similar to the one I wrote above + changes from the link I posted and it, theoretically, could work. It'll take some work, but you may be able to figure something out.

**Comment by mrbarletta** [comment]
Thanks @fdncred for your prompt suggestion! 

It did work! thanks!  

```
{
      name: "fzf_history_menu_fzf_ui_session"
      modifier: none
      keycode: up
      mode: emacs
      event: { send: menu name: fzf_history_menu_fzf_ui_session }
}
```

Used that and it opens up the Menu. 

In your mighty knowledge, do you know how to hijack the default arrow up history  command?  that will be my preferred way instead of this new menu.

thanks again!

**Comment by fdncred** [comment]
> do you know how to hijack the default arrow up history command? that will be my preferred way instead of this new menu.

I'm not sure I understand your quesiton. What you have above hijack's the default up arrow.

You can also do things like this that modifies the command line in the repl with a nushell command.
```
    {
        name: change_dir_with_fzf
        modifier: control
        keycode: char_f
        mode: emacs
        event: {
            send: executehostcommand,
            cmd: "cd (ls | where type == dir | each { |it| $it.name} | str join (char nl) | fzf | decode utf-8 | str trim)"
        }
    }
```
The only parts that I'm not sure about is how to traverse the history using successive up arrows but we'd need to have something closer to this script in place first, just not using it as a menu https://github.com/nushell/nu_scripts/blob/main/custom-menus/current_session_history_menu.nu

**Reply by mrbarletta** [reply]
Thanks for the follow up and the ideas.

I will try to explain it differently.

By default, `arrow up` shows the last command in the history without a menu. 

Lets say I misspelled the last command, with the default history command, I will have to   `arrow up`, ` correct typo`, `enter` . 

Now, with the menu that shows up with the `arrow up` keycode I will have to type a bit more, it will be `arrow up`, `enter`,  `correct typo`, `enter`. 

Thanks!
